### PR TITLE
Add frontend serverSecurity service

### LINKED ISSUE #6959: Extract common services from main plugin
## Description
Historically the main Wazuh dashboard plugin contains many common services needed for the different modules to work. We are in the process of segregating the different modules into decoupled plugins, so we need a common service provider for all of them. Therefore we need to evaluate which services are common to the modules and extract them from main plugin into a common service provider plugin.

#### Objective 
- https://github.com/wazuh/wazuh/issues/22887

## Tasks
The steps that have to be completed in order to close the issue.

- [x] Identify and make a list of the services to extract https://github.com/wazuh/wazuh-dashboard-plugins/issues/6959#issuecomment-2345873062
- [ ] Extract the services into a common provider
  - [x] https://github.com/wazuh/wazuh-dashboard-plugins/pull/7000
  - [x] https://github.com/wazuh/wazuh-dashboard-plugins/pull/7026
  - [ ] https://github.com/wazuh/wazuh-dashboard-plugins/pull/7037
- [ ] Implement the common provider in each module

## Functional requirements
- We must user the mechanisms provided in the life cycle of OpenSearch plugins to expose them, so other plugins can make use of them.

### PR DESCRIPTION
**Pull request by Desvelao**
### Description
This pull request adds a `serverSecurity` client to the `wazuh-core` plugin and expose it to manage the security of the Wazuh server is a composition of old services:
- WzUserPermissions
- useUserPermissions
- useUserPermissionsRequirements
- withUserAuthorizationPrompt
- withUserLogged
- withUserPermissions
- withUserPermissionsRequirements
- WzButtonPermissions
- WzElementPermissions

Changes:
- Refactor the loading while getting the server API token using a similar render than the platform
<details>
  <summary>Screenshot</summary>
  
  image
</details>

- Add documentation for `dashboardSecurity` service
- Move server user authentication data to `dashboardSecurity` and `serverSecurity` services
- Add hooks and HOCs to `dashboardSecurity` service
 
### Issues Resolved
#6959

### Check List
- [x] All tests pass
  - [x] `yarn test:jest`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: plugins/wazuh-core/public/services/server-security/ui/hocs/creator.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  useServerUserPermissionsRequirements,
+  useServerUserPermissionsIsAdminRequirements,
+  useServerUserLogged,
+  useLoadingLogo,
+  PromptNoPermissions,
+  LoadingServerUserLogging,
+}) => {
+  const withServerUserAuthorizationPromptChanged =
+    (permissions = null, othersPermissions = { isAdmininistrator: null }) =>
+    WrappedComponent =>
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/hocs/creator.tsx:14]:
```suggestion
    (WrappedComponent: React.ElementType) =>
```

File: plugins/wazuh-core/public/services/server-security/ui/hocs/creator.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        <PromptNoPermissions
+          permissions={userPermissionRequirements}
+          administrator={userPermissionIsAdminRequirements}
+        />
+      ) : (
+        <WrappedComponent {...props} />
+      );
+    };
+
+  const withServerUserLogged = WrappedComponent => props => {
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/hocs/creator.tsx:38]:
```suggestion
  const withServerUserLogged = (WrappedComponent: React.ElementType) => props => {
```

File: plugins/wazuh-core/public/services/server-security/ui/hocs/creator.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    return withServerUserLogged ? (
+      <WrappedComponent {...props} />
+    ) : (
+      <LoadingServerUserLogging useLoadingLogo={useLoadingLogo} />
+    );
+  };
+
+  const withServerUserAuthorizationPrompt =
+    (permissions = null, othersPermissions = { isAdmininistrator: null }) =>
+    WrappedComponent =>
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/hocs/creator.tsx:50]:
```suggestion
    (WrappedComponent: React.ElementType) =>
```

File: plugins/wazuh-core/public/services/server-security/ui/hocs/creator.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+export const createServerSecurityHOCS = ({
+  useServerUserPermissionsRequirements,
+  useServerUserPermissionsIsAdminRequirements,
+  useServerUserLogged,
+  useLoadingLogo,
+  PromptNoPermissions,
+  LoadingServerUserLogging,
+}) => {
+  const withServerUserAuthorizationPromptChanged =
+    (permissions = null, othersPermissions = { isAdmininistrator: null }) =>
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/hocs/creator.tsx:13]:
```suggestion
    (permissions: Function | object | null = null, othersPermissions = { isAdmininistrator: null }) =>
```

File: plugins/wazuh-core/public/services/server-security/ui/hocs/creator.tsx (hunk 1 above)
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/hocs/creator.tsx:14]:
```suggestion
    (WrappedComponent: React.ElementType) =>
```

File: plugins/wazuh-core/public/services/server-security/ui/components/format.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+              - {PermissionFormatter(permission)}
+            </div>
+          );
+        }
+      })}
+    </div>
+  );
+};
+
+const PermissionFormatter = (permission, key?: string) =>
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/components/format.tsx:55]: Define before use it

File: plugins/wazuh-core/public/services/server-security/ui/components/format.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    <Fragment {...(key ? { key } : {})}>
+      <strong>{permission.action}</strong> (
+      <span style={{ textDecoration: 'underline' }}>{permission.resource}</span>
+      )
+    </Fragment>
+  ) : (
+    <strong {...(key ? { key } : {})}>{permission}</strong>
+  );
+
+const getPermissionComponentKey = permission =>
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/components/format.tsx:66]: Define before use it

File: plugins/wazuh-core/public/services/server-security/ui/components/button.tsx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+      : EuiButton;
+
+  return (
+    <ServerElementPermissions
+      permissions={permissions}
+      administrator={administrator}
+      tooltip={tooltip}
+      getAdditionalProps={disabled => {
+        const additionalProps = {
+          ...(!['link', 'switch'].includes(buttonType)
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/components/button.tsx:62]: Use the enum created in the previous comment.

File: plugins/wazuh-core/public/services/server-security/ui/components/button.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            ? { isDisabled: disabled }
+            : { disabled }),
+          onClick: disabled || !rest.onClick ? undefined : rest.onClick,
+          onChange:
+            !disabled || rest.onChange || buttonType === 'switch'
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/components/button.tsx:67]: Use the enum created in the previous comment.

File: plugins/wazuh-core/public/services/server-security/ui/components/button.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+              ? rest.onChange
+              : undefined,
+        };
+        if (buttonType == 'switch') delete additionalProps.onClick;
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/components/button.tsx:71]: Use the enum created in the previous comment.

File: plugins/wazuh-core/public/services/server-security/ui/components/element.tsx [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  tooltip?: any;
+  children: React.ReactElement;
+  getAdditionalProps?: (disabled: boolean) => {
+    [prop: string]: any;
+  };
+}
+
+export const ServerElementPermissions = ({
+  children,
+  permissions = null,
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/ui/components/element.tsx:38]:
```suggestion
  permissions: Function | object | null = null,
```

File: plugins/wazuh-core/public/services/server-security/wz-user-permissions.ts [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    requiredRole => !userRoles.includes(requiredRole),
+  );
+  return rolesUserNotOwn.length ? rolesUserNotOwn : false;
+};
+
+// Constants
+const RESOURCE_ANY = '*:*:*';
+const RESOURCE_ANY_SHORT = '*:*';
+const RBAC_MODE_BLACK = 'black';
+const RBAC_MODE_WHITE = 'white';
```
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/wz-user-permissions.ts]: Fixed here: https://github.com/wazuh/wazuh-dashboard-plugins/pull/7026/commits/e2dd1be748666eb9b485aa273ccaa2e71b4107bf

File: plugins/wazuh-core/public/services/server-security/ui/components/format.tsx [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+                {permission
+                  .map(p =>
+                    PermissionFormatter(
+                      p,
+                      `no-permissions-${getPermissionComponentKey(
+                        permission,
+                      )}-${getPermissionComponentKey(p)}`,
+                    ),
+                  )
+                  .reduce((prev, cur) => [prev, ', ', cur])}
```
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/format.tsx:36]: This does a "join" of ReactJS components.

File: plugins/wazuh-core/public/services/server-security/ui/components/button.tsx (hunk 9 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/button.tsx]: Fixed here: https://github.com/wazuh/wazuh-dashboard-plugins/pull/7026/commits/bb3ca50c7a90c743c1ac633c508103e6d3eb3837

File: plugins/wazuh-core/public/services/server-security/ui/components/button.tsx (hunk 7 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/button.tsx]: Fixed here: https://github.com/wazuh/wazuh-dashboard-plugins/pull/7026/commits/bb3ca50c7a90c743c1ac633c508103e6d3eb3837

File: plugins/wazuh-core/public/services/server-security/ui/components/button.tsx [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+  ...rest
+}: IServerButtonPermissionsProps) => {
+  const Button =
+    buttonType === 'empty'
+      ? EuiButtonEmpty
+      : buttonType === 'icon'
+      ? EuiButtonIcon
+      : buttonType === 'link'
+      ? EuiLink
+      : buttonType === 'switch'
```
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/button.tsx]: Fixed here: https://github.com/wazuh/wazuh-dashboard-plugins/pull/7026/commits/bb3ca50c7a90c743c1ac633c508103e6d3eb3837

File: plugins/wazuh-core/public/services/server-security/ui/components/button.tsx (hunk 8 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/button.tsx]: Fixed here: https://github.com/wazuh/wazuh-dashboard-plugins/pull/7026/commits/bb3ca50c7a90c743c1ac633c508103e6d3eb3837

File: plugins/wazuh-core/public/services/server-security/ui/components/format.tsx (hunk 5 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/format.tsx]: Fixed here: https://github.com/wazuh/wazuh-dashboard-plugins/pull/7026/commits/1ada4ae8e6559e58c2d0de19f06489eed87cba71

File: plugins/wazuh-core/public/services/server-security/ui/components/format.tsx (hunk 6 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/format.tsx]: Fixed here: https://github.com/wazuh/wazuh-dashboard-plugins/pull/7026/commits/1ada4ae8e6559e58c2d0de19f06489eed87cba71

File: plugins/wazuh-core/public/services/server-security/ui/components/element.tsx (hunk 10 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/element.tsx:38]: The type for `permissions` are defined here https://github.com/wazuh/wazuh-dashboard-plugins/pull/7026/files#diff-db1ea29a2b610976d178ef552064997488c67d5efde4841c737f533039dbdd6eR45

File: plugins/wazuh-core/public/services/server-security/wz-user-permissions.ts [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+// Data extrated of /security/actions endpoint
+import wazuhPermissions from '../../../common/api-info/security-actions.json';
+
+// Constants
+const RESOURCE_ANY = '*:*:*';
+const RESOURCE_ANY_SHORT = '*:*';
+const RBAC_MODE_WHITE = 'white';
+// Utility functions
+const isAllow = (value: string) => value === 'allow';
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/wz-user-permissions.ts:21]: Replace with a constant instead (`allow`)

File: plugins/wazuh-core/public/services/server-security/types.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    useServerUserPermissions: () => any;
+    useServerUserPermissionsRequirements: (
+      permissions: ServerSecurityCombinedPermissionWithFunction,
+    ) => [ServerSecurityCombinedPermission, any];
+  };
+  hocs: {
+    withServerUserAuthorizationPrompt: (
+      permissions: ServerSecurityCombinedPermissionWithFunction | null,
+      otherPermissions: { isAdmininistrator: boolean | null },
+    ) => (WrappedComponent: React.Component) => React.ReactElement;
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/types.ts:37]:
```suggestion
    ) => (WrappedComponent: React.ElementType) => React.ReactElement;
```

File: plugins/wazuh-core/public/services/server-security/types.ts [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    withServerUserLogged: (
+      WrappedComponent: React.Component,
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/server-security/types.ts:39]:
```suggestion
      WrappedComponent: React.ElementType,
```

**Review by guidomodarelli — CHANGES_REQUESTED** [review]

File: plugins/wazuh-core/public/services/server-security/ui/components/format.tsx [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+  | ServerSecurityPermissionRequirement[]
+  | ServerSecurityPermission
+  | ServerSecurityPermissionAsString;
+
+const PermissionFormatter = (
+  permission: ServerSecurityPermissionRequirement,
+  key?: string,
+) =>
+  typeof permission === 'object' ? (
+    <Fragment {...(key ? { key } : {})}>
```
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/format.tsx:37]: This generates a new key in each render and should use the same key if the element is the same.

The source code of the component was ported from `main` plugin.

File: plugins/wazuh-core/public/services/server-security/ui/components/format.tsx [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+      <strong>{permission.action}</strong> (
+      <span style={{ textDecoration: 'underline' }}>{permission.resource}</span>
+      )
+    </Fragment>
+  ) : (
+    <strong {...(key ? { key } : {})}>{permission}</strong>
```
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/format.tsx:43]: This generates a new key in each render and should use the same key if the element is the same.

The source code of the component was ported from main plugin.

File: plugins/wazuh-core/public/services/server-security/ui/components/button.tsx [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+          ].includes(buttonType)
+            ? { disabled }
+            : { isDisabled: disabled }),
+          onClick: disabled || !rest.onClick ? undefined : rest.onClick,
+          onChange:
+            !disabled ||
+            rest.onChange ||
+            buttonType === SERVER_BUTTON_PERMISSIONS_TYPES.SWTICH
+              ? rest.onChange
+              : undefined,
```
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/button.tsx:79]: What do you mean?

File: plugins/wazuh-core/public/services/server-security/wz-user-permissions.ts (hunk 14 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/wz-user-permissions.ts]: Solved in 12187ab1638412e0cc8155aecfd00cbe75a6253f

File: plugins/wazuh-core/public/services/server-security/types.ts (hunk 15 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/types.ts]: Solved in 12187ab1638412e0cc8155aecfd00cbe75a6253f.

Use `React.ComponentType` instead.

File: plugins/wazuh-core/public/services/server-security/types.ts (hunk 16 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/types.ts]: Solved in 12187ab1638412e0cc8155aecfd00cbe75a6253f.

Use `React.ComponentType` instead.

File: plugins/wazuh-core/public/services/dashboard-security/dashboard-security.ts [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+    try {
+      this.logger.debug('Fetching the security platform');
+
+      const response = await this.http.get(
+        '/elastic/security/current-platform',
+      );
+
+      this.logger.debug(`Security platform: ${this.securityPlatform}`);
+
+      this.securityPlatform = response.platform;
```
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/wazuh-core/public/services/dashboard-security/dashboard-security.ts]: Fixed here 1cd615404f16c6de8a060705de8229f16677f2e2

File: plugins/wazuh-core/public/services/dashboard-security/dashboard-security.ts [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+      this.logger.debug('Created the UI utilities');
+    } catch (error) {
+      this.logger.error(`Error creating the UI utilities: ${error.message}`);
+      throw error;
+    }
+
+    try {
+      this.logger.debug('Getting security platform');
+      this.securityPlatform = await this.fetchCurrentPlatform();
+      this.logger.debug(`Security platform: ${this.securityPlatform}`);
```
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/wazuh-core/public/services/dashboard-security/dashboard-security.ts]: Fixed here 1cd615404f16c6de8a060705de8229f16677f2e2

File: plugins/wazuh-core/public/services/dashboard-security/README.md [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+
+- HOC
+
+```ts
+const MyComponent = getWazuhCorePlugin().hocs.withDashboardSecurityAccountAdmin(
+  ({ dashboardSecurityAccountAdmin }) => {
+    // dashboardSecurityAccountAdmin contains if the user is admin or not
+  },
+);
+```
```
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/wazuh-core/public/services/dashboard-security/README.md]: Enhanced here: a2e5df561f3b3aa4d04ab1e18f917f00f56a6dc2.

I am not sure if this solves the suggestion.

File: plugins/wazuh-core/public/services/server-security/ui/components/button.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/wazuh-core/public/services/server-security/ui/components/button.tsx]: Solved in a meeting.

**Review by guidomodarelli — APPROVED** [review]
LGTM

**Review by yenienserrano — APPROVED** [review]

**Review by guidomodarelli — APPROVED** [review]
LGTM

**Pull request merged**